docker_cli/image.py

```diff
@@ -0,0 +1,430 @@
+import io
+import os
+from shlex import split
+from subprocess import PIPE, STDOUT, CalledProcessError, run
+from sys import stdin
+from typing import (Any, List, Optional, Sequence, Type, TypeVar, Union,
+                    overload)
+
+from .exceptions import CommandNotFoundOnImageError
+
+
+class Image:
+    """A class which holds a reference to a given docker image and provides
+    an interface by which to interact with that image.
+
+    -   Capable of building docker images from dockerfiles or dockerfile-
+        formatted strings via the build classmethods.
+    -   Capable of acquiring information about the given image (see property
+        methods below).
+    -   Capable of running commands on containers built from the image using
+        the run method.
+    -   All docker build and run methods on the class can recive a file-like
+        object, (io.TextIOBase or some subclass of it,) and they will write the
+        outputs from docker into the file to enable logging.
```

Comment:
Few minor suggestions here:

* I like that we're summarizing the main capabilities of the class here, but I think I would skip the details like the ability to log command output to a file -- I'd save that for the individual method docstrings below.

* In addition, I'd suggest prefixing this list with something like "Capabilities include: " in order to avoid repeating "Cabable of " in each bullet.

* Finally, if you want to reference individual methods like `build()` or `run()` from this docstring, let's use [sphinx's cross-referencing syntax](https://stackoverflow.com/a/21297186) so that a link to the method could be inserted if we rendered html pages from these docstrings.

```suggestion
    Capabilities include:

    -   Building docker images from dockerfiles or dockerfile-formatted strings
        via :func:`~docker_cli.Image.build`.
    -   Running commands in containers built from the image using
        :func:`~docker_cli.Image.run`.
    -   Inspecting properties of the given image.
```